### ISSUE TITLE
Parameterized params of func pointer params cannot have broader types than the bound type

### ISSUE THREAD

**Issue by pubudu91** [ballerina-platform/ballerina-lang#16759]
Consider the following:
```ballerina
function testForeachOpWithClosedRecords() returns map<any> {
    ClosedPerson p = { name: "John Doe", age: 25, address: { street: "Palm Grove", city: "Colombo 3" }};
    map<any> rec = {};
    int fName = 0;

    p.forEach(function (any value) { // `any` given for the type param
        rec[fName.toString()] = value;
        fName += 1;
    });

    return rec;
}
```
The above gives the following error:
```
incompatible types: expected '(string|int|Address)', found 'any'
```

**Comment by MaryamZi** [comment]
We currently allow having narrower types instead, which may lead to class cast exceptions.

For example
```ballerina
import ballerina/io;

function testForeachOpWithClosedRecords() returns map<any> {
    ClosedPerson p = { name: "John Doe", age: 25, 
                                    address: { street: "Palm Grove", city: "Colombo 3" }};
    map<any> rec = {};
    int fName = 0;

    p.forEach(function (string value) { // `string` given for the type param
        rec[fName.toString()] = value;
        fName += 1;
    });

    return rec;
}

type ClosedPerson record {|
    string name;
    int age;
    map<anydata> address;
|};

public function main() {
    io:println(testForeachOpWithClosedRecords());
}
```

```cmd
$ ballerina run err_test.bal
java.lang.ClassCastException: java.lang.Long cannot be cast to java.lang.String
        at err_test.$lambda$0$lambda$(err_test.bal)
        at org.ballerinalang.jvm.values.FPValue.accept(FPValue.java:59)
        at org.ballerinalang.langlib.map.ForEach.lambda$forEach$0(ForEach.java:41)
        at java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
        at org.ballerinalang.langlib.map.ForEach.forEach(ForEach.java:41)
        at ballerina.lang_map.map.forEach(map.bal:64)
        at err_test.testForeachOpWithClosedRecords(err_test.bal:8)
        at err_test.main(err_test.bal:23)
        at ___init.$lambda$main$(.)
        at org.ballerinalang.jvm.SchedulerItem.execute(Scheduler.java:453)
        at org.ballerinalang.jvm.Scheduler.run(Scheduler.java:216)
        at org.ballerinalang.jvm.Scheduler.runSafely(Scheduler.java:187)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by MaryamZi** [comment]
Fixed with #17062.
